### PR TITLE
Optimize SelectRange typeclass derivation

### PR DESCRIPTION
**Pull request by Ssstlis**
Hi all!
Once i've find out the SelectRange is useful, but i don't understand why is this instances are so hard to derive (too much unnecessary recursive calling).
So i decide to optimize this derivation and it still works and looks good.

### CODE REVIEW

**Comment by milessabin** [comment]
@joroKr21 I'd appreciate your eyeballs on this :-)

File: core/src/main/scala/shapeless/ops/hlists.scala [hunk 1]
```diff
@@ -946,12 +946,22 @@ object hlist {
 
     type Aux[L <: HList, A <: Nat, B <: Nat, Out0 <: HList] = SelectRange[L, A, B] {type Out = Out0}
 
-    implicit def SelectRangeAux[L <: HList, A <: Nat, B <: Nat, Ids <: HList, SelOut <: HList]
-    (implicit range: shapeless.ops.nat.Range.Aux[A, B, Ids], sel: SelectMany.Aux[L, Ids, SelOut]): Aux[L, A, B, SelOut] =
-      new SelectRange[L, A, B] {
-        type Out = SelOut
+    implicit def thenTake[L <: HList, T <: Nat](
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/ops/hlists.scala]: Not sure about the name of this method. Why not just call it `take`?

**Comment by joroKr21** [comment]
LGTM but I'm not sure about the naming.

File: core/src/main/scala/shapeless/ops/hlists.scala (hunk 1 above)
Comment (Ssstlis) [reply core/src/main/scala/shapeless/ops/hlists.scala]: Then i should rename also a param name. Ok, i have an idea.

**Comment by joroKr21** [comment]
@Ssstlis one more thing before merge, would you mind squashing all commits into one?

**Comment by Ssstlis** [comment]
squashed and rebased

**Pull request merged**